### PR TITLE
Compare new and existing processed data

### PR DESCRIPTION
**Pull request by hacodeorg**
Compares content in `contact_rollups_processed` table and the existing `contact_rollups_final` table (which is an older version of contact_rollups_processed from previous pipeline execution). 

The results are saved into `contact_rollups_comparisons` table. Data in this table will be used later to determine what we should send to Pardot.

This PR is based on the migrations in https://github.com/code-dot-org/code-dot-org/pull/33677

## Todo
- [ ] Test on adhoc

## Links
* [PLC-810](https://codedotorg.atlassian.net/browse/PLC-810)
* [CRv1.5 table schemas](https://docs.google.com/document/d/1dMO8-cGQjtX1jZ32XIfPuk5Y20dYU_dSU8_7i7Vzcp8/edit#heading=h.yxvhor9o8zb8)

## Testing story
* Unit tested
* Will test with the clone of production DB via adhoc machine.

# Reviewer Checklist:
- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/contact_rollups_comparison.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    SQL
+
+    # Saves results of the two queries above
+    insert_records_query = <<-SQL.squish
+      INSERT INTO #{ContactRollupsComparison.table_name}
+        (email, new_data, new_data_updated_at, old_data, old_data_updated_at, created_at)
+      #{find_existing_records_query}
+      UNION ALL
+      #{find_new_records_query}
+    SQL
```
Comment (islemaster) [inline dashboard/app/models/contact_rollups_comparison.rb:48]: I see you, simulated `FULL OUTER JOIN`.  (Maybe worth a comment to that effect?)

**Review by islemaster — APPROVED** [review]
:+1:

File: dashboard/app/models/contact_rollups_comparison.rb (hunk 1 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_comparison.rb:48]: Yes, that's exactly what it is. Will add a comment :)

File: dashboard/app/models/contact_rollups_comparison.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      FROM contact_rollups_final AS old_data
+      RIGHT OUTER JOIN contact_rollups_processed AS new_data
+      ON old_data.email = new_data.email
+      WHERE old_data.email IS NULL
+    SQL
+
+    # Saves results of the two queries above
+    insert_records_query = <<-SQL.squish
+      INSERT INTO #{ContactRollupsComparison.table_name}
+        (email, new_data, new_data_updated_at, old_data, old_data_updated_at, created_at)
```
Comment (hacodeorg) [inline dashboard/app/models/contact_rollups_comparison.rb]: The column order here is wrong, should be `email, old_data, old_data_updated_at, new_data, new_data_updated_at, created_at`. Will fix.

File: dashboard/app/models/contact_rollups_comparison.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    find_existing_records_query = <<-SQL.squish
+      SELECT old.email, old.data, old.updated_at, new.data, new.updated_at, now()
+      FROM contact_rollups_final AS old
+      LEFT OUTER JOIN contact_rollups_processed AS new
+      ON old.email = new.email
+    SQL
+
+    # Finds all records in the new data but not in the old data
+    find_new_records_query = <<-SQL.squish
+      SELECT new.email, old.data, old.updated_at, new.data, new.updated_at, now()
```
Comment (bencodeorg) [inline dashboard/app/models/contact_rollups_comparison.rb:34]: Might be nice to be explicit here and say `null as old_data, null as old_updated_at` to show that's what you expect in those columns.

File: dashboard/test/models/contact_rollups_comparison_test.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    ContactRollupsComparison.compile_processed_data
+
+    assert_equal new_contact_count, ContactRollupsComparison.count
+  end
+
+  test 'compile_processed_data when new data is empty' do
+    clean_tables
+    old_contact_count = 3
+    create_list :contact_rollups_processed, old_contact_count
```
Comment (bencodeorg) [inline dashboard/test/models/contact_rollups_comparison_test.rb]: Shouldn't this create instances of `contact_rollups_final`?

File: dashboard/test/models/contact_rollups_comparison_test.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    assert_equal old_contact_count, ContactRollupsComparison.count
+  end
+
+  test 'compile_processed_data when new and old data overlap' do
+    clean_tables
+
+    total_contact_count = 9
+    (0..5).each {|index| create :contact_rollups_processed, email: "email#{index}@example.domain"}
+    (3..total_contact_count - 1).each {|index| create :contact_rollups_final, email: "email#{index}@example.domain"}
```
Comment (bencodeorg) [inline dashboard/test/models/contact_rollups_comparison_test.rb]: Nice!

**Review by bencodeorg — COMMENTED** [review]
I think there's a test fix to make but otherwise looks good!

File: dashboard/test/models/contact_rollups_comparison_test.rb (hunk 4 above)
Comment (hacodeorg) [reply dashboard/test/models/contact_rollups_comparison_test.rb]: Yes I just fixed it.

File: dashboard/app/models/contact_rollups_comparison.rb (hunk 3 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_comparison.rb:34]: These only purpose of these 2 queries is to simulate a full outer join. I'd prefer not to add other logic on top of it.

**Review by bencodeorg — APPROVED** [review]

**Pull request merged**